### PR TITLE
add prefix to bundled hwloc symbols

### PR DESCRIPTION
**Pull request by jhh67**
Add unique prefix to bundled hwloc symbols. The bundled hwloc is curr…ently v1, which can cause a
version mismatch with third-party software such as libopen-pal that uses the v2 version of hwloc.
Adding a prefix to the bundled hwloc symbols ensures that only software that uses the prefix will
use the bundled hwloc.

This change involves adding a --with-hwloc-symbol-prefix configuration option to hwloc. Note that
hwloc/rename.h creates macros for the original hwloc symbol names so that neither hwloc nor the
software that uses it needs to be modified to use the new symbol names. For Chapel this includes
the runtime source as well as the third-party qthreads library. However, qthread configuration
fails if hwloc provides non-standard symbol names, requiring a change to qthread to
specify the --with-hwloc-symbol-prefix prefix when configuring it to use the bundled hwloc.

### CODE REVIEW

**Comment by gbtitus** [comment]
@ronawho What do you think is the best way to address the autogen-related changes? I don't think we want to push those in an upstream change, definitely, and probably even for our own uses it would be better to re-gen using the same autogen tools versions the Qthreads folks did. But how best to achieve that?

**Comment by gbtitus** [comment]
There's a funny '...' in the Description block for the PR, maybe from cut+paste from a Mac? (I've seen oddball characters get pasted that way before.)

**Comment by ronawho** [comment]
> @ronawho What do you think is the best way to address the autogen-related changes? I don't think we want to push those in an upstream change, definitely, and probably even for our own uses it would be better to re-gen using the same autogen tools versions the Qthreads folks did. But how best to achieve that?

When I pull in gasnet changes or other third-party libs that require a regen I use the same version the official release did. I typically build on chapcs and use spack to install the same version.

File: third-party/hwloc/Makefile.include
Comment (gbtitus) [inline third-party/hwloc/Makefile.include:11]: I may be missing something -- is it not possible just to define `CHPL_HWLOC_PREFIX=chpl_` (or nothing) and throw `--with-hwloc-symbol-prefix=$CHPL_HWLOC_PREFIX` to both the hwloc and qthreads `configure` scripts? Then default the hwloc prefix string to nothing and do `AC_SEARCH_LIBS([${with_hwloc_symbol_prefix}hwloc_topology_init], ...`in qthread_check_hwloc.m4?

File: third-party/qthread/Makefile
Comment (gbtitus) [inline third-party/qthread/Makefile:136]: Maybe check for nonblank `CHPL_HWLOC_PREFIX` here instead of bundled hwloc? Seems more direct.

File: third-party/qthread/Makefile
Comment (ronawho) [inline third-party/qthread/Makefile:137]: I'd throw this up next to `CHPL_QTHREAD_CFG_OPTIONS += --with-hwloc=$(HWLOC_INSTALL_DIR)`

**Comment by jhh67** [comment]
> > @ronawho What do you think is the best way to address the autogen-related changes? I don't think we want to push those in an upstream change, definitely, and probably even for our own uses it would be better to re-gen using the same autogen tools versions the Qthreads folks did. But how best to achieve that?
> 
> When I pull in gasnet changes or other third-party libs that require a regen I use the same version the official release did. I typically build on chapcs and use spack to install the same version.

I'm not familiar with spack, where do I find documentation?

**Comment by ronawho** [comment]
> I'm not familiar with spack, where do I find documentation?

I added instructions to my initial comment.

File: third-party/hwloc/Makefile.include
Comment (jhh67) [reply third-party/hwloc/Makefile.include:11]: I think we want CHPL_HWLOC_PREFIX to be "chpl_hwloc_" to be consistent with the value of CHPL_JEMALLOC_PREFIX (chpl_je_). The problem is if you that value of CHPL_HWLOC_PREFIX to  --with-hwloc-symbol-prefix for hwloc you end up with the prefix being "chpl_hwloc_hwloc_" (this is essentially a bug on their part because it doesn't work the way it's documented or the way other packages handle prefixes). You can't pass the same prefix to hwloc and qthread, so I hid the ugliness in the hwloc Makefile.

File: third-party/qthread/Makefile
Comment (jhh67) [reply third-party/qthread/Makefile:136]: Somewhere I have to check for bundled hwloc to know whether or not to set the CHPL_HWLOC_PREFIX. I could do that in hwloc/Makefile.in if you think that's a better solution.

**Comment by jhh67** [comment]
> > I'm not familiar with spack, where do I find documentation?
> 
> I added instructions to my initial comment.

Thanks, I'll check it out.

**Comment by ronawho** [comment]
I don't know that we've documented this anywhere, but we typically try to `git am` in the third-party change (to maintain committer history, though in this case that's you so it's not super important) and include the modification in the README. Updating the README serves 2 purposes -- tracking what changes we have from the official third-party, and our build system keys off that to detect that we need to rebuild. 

See https://github.com/chapel-lang/chapel/pull/17083 for a recent example in gasnet.

**Comment by jhh67** [comment]
> > @ronawho What do you think is the best way to address the autogen-related changes? I don't think we want to push those in an upstream change, definitely, and probably even for our own uses it would be better to re-gen using the same autogen tools versions the Qthreads folks did. But how best to achieve that?
> 
> When I pull in gasnet changes or other third-party libs that require a regen I use the same version the official release did. I typically build on chapcs and use spack to install the same version.
> 
> e.g.:
> 
> ```shell
> git clone https://github.com/spack/spack.git
> cd spack
> source share/spack/setup-env.sh
> spack install automake@1.13.4
> spack load automake@1.13.4
> ```

Also, is it best to put all the autogenerated files in a separate commit?

File: third-party/qthread/Makefile
Comment (gbtitus) [reply third-party/qthread/Makefile:136]: If we're not using a bundled hwloc then `CHPL_HWLOC_PREFIX` will not be set and `$(CHPL_HWLOC_PREFIX)` will just be empty. But to make it more obvious we could do `ifndef CHPL_HWLOC_PREFIX` instead of checking for an unset value.

**Comment by ronawho** [comment]
> Also, is it best to put all the autogenerated files in a separate commit?

Yeah, I think so

File: third-party/hwloc/Makefile.include
Comment (gbtitus) [reply third-party/hwloc/Makefile.include:11]: I'd defer to @ronawho on this one, too, but personally I don't think we have to try to make the hwloc prefix look like it works the same as the jemalloc one even though it doesn't. In fact, making it more obvious that it's different seems better, to me. Since Qthreads hasn't previously had support for a hwloc prefix, aren't we in control of whether the hwloc symbol prefix specified to the Qthreads configure script needs to include the `hwloc_` part or not? It seems as though we could just make the Qthreads behavior with respect to the hwloc prefix match the hwloc behavior, even though neither matches the jemalloc behavior.

File: third-party/qthread/qthread-src/configure
Comment (gbtitus) [inline third-party/qthread/qthread-src/configure:3250]: `configure` is a generated file (from `configure.ac`). Are these changes unrelated to hwloc due to using a different `autoconf` version that the one the Qthreads team uses? And should the hwloc-related changes in it actually be made to `configure.ac` instead?

**Comment by gbtitus** [comment]
I like the new hwloc prefix logic!

**Review by gbtitus — APPROVED** [review]
👍  Fab!

**Review by ronawho — APPROVED** [review]

**Pull request merged**
